File: phase B/cs240_main_phaseB_v13122017/Java/Main.java
/**********************************************************************
 * @file   Main.java                                                  *
 * @author Nikolaos Batsaras					      *
 *                                                                    *
 * @brief  Main file for the needs of the Data Structures (HY-240b)   *
 * project (Fall 2017)                                                *
 * Computer Science Department, University of Crete, Greece           *
 **********************************************************************/

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;

public class Main {

    static int max_soldiers_g;
    
    /*
     * Here you need to declare:
     * 1) registration_hashtable
     * 2) generals_list
     * 3) my_combat
     *
     * The classes related to these variables will be created by you.
     * After your classes are ready, declare the above variables.
     * /

    /**
     * @brief Optional function to initialize data structures that 
     * need initialization
     *
     * @return true on success
     *         false on failure
     */
    public static boolean initialize() {
        return true;
    }

    /**
     * @brief Register soldier
     *
     * @param sid The soldier's id
     * @param gid The general's id who orders the soldier
     *
     * @return true on success
     *         false on failure
     */
    public static boolean register(int sid, int gid) {
        return true;
    }
    /**
     * @brief General or King joins the battlefield
     *
     * @param gid The general's id
     * @return true on success
     *         false on failure
     */
    public static boolean add_general(int gid) {
        return true;
    }
    /**
     * @brief Distribute soldiers to the camps
     * 
     * @return true on success
     *         false on failure
     */
    public static boolean distribute() {
        return true;
    }

    /**
     * @brief General resigns from battlefield
     *
     * @param gid1 The id of general that resigns
     * @param gid2 The id of the general that will  
     * 
     * @return true on success
     *         false on failure
     */
    public static boolean general_resign(int gid1, int gid2) {
        return true;
    }

    /**
     * @brief Prepare for battle
     * 
     * @param gid1 The id of the first general
     * @param gid2 The id of the second general
     * 
     * @return true on success
     *         false on failure
     */
    public static boolean prepare_battle(int gid1, int gid2) {
        return true;
    }

    /**
     * @brief Battle
     * 
     * @param god_favor 
     * @param bit_stream A sequence of bits
     * @param bitsno The number of bits in the sequence
     *
     * @return true on success
     *         false on failure
     */
    public static boolean battle(int god_favor, String bit_stream, int bitsno) {
        return true;
    }

    /**
     * @brief Cease fire
     * 
     * @return true on success
     *         false on failure
     */
    public static boolean cease_fire() {
        return true;
    }

    /**
     * @brief Print general's soldiers
     *
     * @param gid The general's id
     *
     * @return true on success
     *         false on failure
     */
    public static boolean print_generals_soldiers(int gid) {
        return true;
    }

    /**
     * @brief Print generals
     *
     * @return true on success
     *         false on failure
     */
    public static boolean print_generals() {
        return true;
    }

    /**
     * @brief Print registration hashtable
     *
     * @return true on success
     *         false on failure
     */
    public static boolean print_registration_hashtable() {
        return true;
    }



    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) throws FileNotFoundException, IOException {
        BufferedReader inputFile;
        String line;
        String [] params;

        /* Check command buff arguments */       
        if (args.length != 1) {
            System.err.println("Usage: <executable-name> <input_file>");
            System.exit(0);
        }

        /* Open input file */        
        inputFile = new BufferedReader(new FileReader(args[0]));

        /* Read input file and handle the events */
        while ( (line = inputFile.readLine()) != null ) {

            if (line.length() == 0) continue;

            System.out.println(">>> Event: " + line);
            params = line.split(" ");
            char eventID = line.charAt(0);

            switch(eventID) {

                /* Comment */
                case '#':
                    break;

                    /* Maximum soldiers */
                case '0':
                    {
                        max_soldiers_g = Integer.parseInt(params[1]);

                        initialize();

                        break;
                    }

                /* Register soldier
                 * R <sid> <gid> */
                case 'R':
                    {
                        int sid = Integer.parseInt(params[1]);
                        int gid = Integer.parseInt(params[2]);

                        if (register(sid, gid)) {
                            System.out.println("R " + sid + " " + gid + " succeeded");
                        } else {
                            System.err.println("R " + sid + " " + gid + " failed");
                        }
                        break;
                    }

                    /* General or king joins the battlefield
                     * G <gid> */
                case 'G':
                    {
                        int gid = Integer.parseInt(params[1]);

                        if (add_general(gid)) {
                            System.out.println("G " + gid + " succeeded");
                        } else {
                            System.err.println("G " + gid + " failed");
                        }
                        break;
                    }

                    /* Distribute soldiers
                     * D */
                case 'D':
                    {   
                        if (distribute()) {
                            System.out.println("D " + " succeeded");
                        } else {
                            System.err.println("D " + " failed");
                        }
                        break;
                    }

                    /* General resigns from battle
                     * M <gid1> <gid2> */
                case 'M':
                    {
                        int gid1 = Integer.parseInt(params[1]);
                        int gid2 = Integer.parseInt(params[2]);

                        if (general_resign(gid1, gid2)) {
                            System.out.println("M " + gid1 + " " + gid2 + " succeeded");
                        } else {
                            System.err.println("M " + gid1 + " " + gid2 + " failed");
                        }
                        break;
                    }

                    /* Prepare for battle
                     * P <gid1> <gid2> */
                case 'P':
                    {
                        int gid1 = Integer.parseInt(params[1]);
                        int gid2 = Integer.parseInt(params[2]);

                        if (prepare_battle(gid1, gid2)) {
                            System.out.println("P " + gid1 + " " + gid2 + " succeeded");
                        } else {
                            System.err.println("P " + gid1 + " " + gid2 + " failed");
                        }
                        break;
                    }

                    /* Battle
                     * B <god_favor> <bit_stream> */
                case 'B':
                    {
                        int god_favor = Integer.parseInt(params[1]);
                        String bit_stream = params[2];

                        if (battle(god_favor, bit_stream, 8)) {
                            System.out.println("B " + god_favor + " " + bit_stream + " succeeded");
                        } else {
                            System.err.println("B " + god_favor + " " + bit_stream + " failed");
                        }
                        break;
                    }

                    /* Cease fire
                     * U */
                case 'U':
                    {
                        if (cease_fire()) {
                            System.out.println("U " + " succeeded");
                        } else {
                            System.err.println("U " + " failed");
                        }
                        break;
                    }

                    /* Print general's soldiers
                     * W <gid> */
                case 'W':
                    {                    
                        int gid = Integer.parseInt(params[1]);

                        if (print_generals_soldiers(gid)) {
                            System.out.println("W " + " " + gid + " " + "succeeded");
                        } else {
                            System.err.println("W " + " " + gid + " " + "failed");
                        }
                        break;
                    }

                    /* Print generals
                     * X */
                case 'X':
                    {
                        if (print_generals()) {
                            System.out.println("X succeeded");
                        } else {
                            System.err.println("X failed");
                        }
                        break;

                    }

                    /* Print registration hashtable
                     * Y */
                case 'Y':
                    {
                        if (print_registration_hashtable()) {
                            System.out.println("Y succeeded");
                        } else {
                            System.err.println("Y failed");
                        }
                        break;
                    }

                    /* Empty line */
                case '\n':
                    break;

                    /* Ignore everything else */
                default:
                    System.out.println("Ignoring " + line);
                    break;
            }
        }

    }

}

File: phase B/cs240_main_phaseB_v13122017/C/main.c
/************************************************************//**
 * @file   main.c                                      		*
 * @author Nikolaos Batsaras <batsaras@csd.uoc.gr>  		*
 *                                                    		*
 * @brief Main function for the needs of cs-240b project 2017   *
 ***************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#include "troy.h"

#define BUFFER_SIZE 1024  /**< Maximum length of a line in input file */

/* Uncomment the following line to enable debugging prints 
 * or comment to disable it */
#define DEBUG

#ifdef DEBUG
#define DPRINT(...) fprintf(stderr, __VA_ARGS__);
#else  /* DEBUG */
#define DPRINT(...)
#endif /* DEBUG */

typedef struct c_soldier c_soldier;
typedef struct general general;
typedef struct TREE_soldier TREE_soldier;
typedef struct soldier soldier;
typedef struct combat combat;

/**
 * @brief Optional function to initialize data structures that 
 * need initialization
 *
 * @return 1 on success
 *         0 on failure
 */
int initialize() {
    return 1;
}

/**
 * @brief Register soldier
 *
 * @param sid The soldier's id
 * @param gid The general's id who orders the soldier
 *
 * @return 1 on success
 *         0 on failure
 */
int register_soldier(int sid, int gid) {
    return 1;
}

/**
 * @brief General or King joins the battlefield
 *
 * @param gid The general's id
 *
 * @return 1 on success
 *         0 on failure
 */
int add_general(int gid) {

    TREE_soldier *newnodeTree = (TREE_soldier*) malloc(sizeof(TREE_soldier));
    newnodeTree->sid = -1;
    newnodeTree->p = NULL;
    newnodeTree->rc = NULL;
    newnodeTree->lc = NULL;

    general *newnode = (general*) malloc(sizeof(general));    newnode->gid = gid;
    newnode->combats_no = 0;
    newnode->soldiers_no = 0;
    newnode->soldiers_R = newnodeTree;
    newnode->soldiers_S = newnodeTree;

    if(generals_list == NULL){
        generals_list = newnode;
        newnode->next = NULL;
    }
    else{
        newnode->next = generals_list;
        generals_list = newnode;
    }
    return 1;
}


/**
 * @brief Distribute soldiers to the camps
 * 
 * @return 1 on success
 *         0 on failure
 */
int distribute() {
    return 1;
}

/**
 * @brief General resigns from battlefield
 *
 * @param gid1 The id of general that resigns
 * @param gid2 The id of the general that will  
 * 
 * @return 1 on success
 *         0 on failure
 */
int general_resign(int gid1, int gid2) {
    return 1;
}

/**
 * @brief Prepare for battle
 * 
 * @param gid1 The id of the first general
 * @param gid2 The id of the second general
 * 
 * @return 1 on success
 *         0 on failure
 */
int prepare_battle(int gid1, int gid2) {
    return 1;
}

/**
 * @brief Battle
 * 
 * @param god_favor
 * @param bit_stream A sequence of bits
 * @param bitsno The number of bits in the sequence
 *
 * @return 1 on success
 *         0 on failure
 */
int battle(int god_favor, char *bit_stream, int bitsno) {
    return 1;
}

/**
 * @brief Cease fire
 * 
 * @return 1 on success
 *         0 on failure
 */
int cease_fire() {
    return 1;
}

/**
 * @brief Print general's soldiers
 *
 * @param gid The id of the general
 *
 * @return 1 on success
 *         0 on failure
 */
int print_generals_soldiers(int gid) {
    return 1;
}

/**
 * @brief Print generals
 *
 * @return 1 on success
 *         0 on failure
 */
int print_generals() {
    general *tempGeneral = generals_list;

    printf("\tGENERALS:\n\t");

    while(tempGeneral != NULL){
        printf("%d: ", tempGeneral->gid );
        tempGeneral = tempGeneral->next;
    }

    printf("DONE\n");
    return 1;
}

/**
 * @brief Print registration hashtable
 *
 * @return 1 on success
 *         0 on failure
 */
int print_registration_hashtable() {
    return 1;
}

/**
 * @brief Free resources
 *
 * @return 1 on success
 *         0 on failure
 */
int free_world() {
    return 1;
}


/**
 * @brief The main function
 *
 * @param argc Number of arguments
 * @param argv Argument vector
 *
 * @return 0 on success
 *         1 on failure
 */
int main(int argc, char** argv)
{
    FILE *fin = NULL;
    char buff[BUFFER_SIZE], event;

    /* Check command buff arguments */
    if ( argc != 2 ) {
        fprintf(stderr, "Usage: %s <input_file> \n", argv[0]);
        return EXIT_FAILURE;
    }

    /* Open input file */
    if (( fin = fopen(argv[1], "r") ) == NULL ) {
        fprintf(stderr, "\n Could not open file: %s\n", argv[1]);
        perror("Opening test file\n");
        return EXIT_FAILURE;
    }

    /* Read input file buff-by-buff and handle the events */
    while ( fgets(buff, BUFFER_SIZE, fin) ) {

        DPRINT("\n>>> Event: %s", buff);

        switch(buff[0]) {

            /* Comment */
            case '#':
            break;

                /* Maximum soldiers */
            case '0':
            {
                sscanf(buff, "%c %d", &event, &max_soldiers_g);
                DPRINT("%c %d\n", event, max_soldiers_g);

                initialize();

                break;
            }

                /* Register soldier
                 * R <sid> <gid> */
            case 'R':
            {
                int sid, gid;

                sscanf(buff, "%c %d %d", &event, &sid, &gid);
                DPRINT("%c %d %d\n", event, sid, gid);

                if ( register_soldier(sid, gid) ) {
                    DPRINT("R %d %d succeeded\n", sid, gid);
                } else {
                    fprintf(stderr, "R %d %d failed\n", sid, gid);
                }

                break;
            }

                /* General or king joins the battlefield
                 * G <gid> */
            case 'G':
            {
                int gid;

                sscanf(buff, "%c %d", &event, &gid);
                DPRINT("%c %d\n", event, gid);

                if ( add_general(gid) ) {
                    DPRINT("%c %d succeeded\n", event, gid);
                } else {
                    fprintf(stderr, "%c %d failed\n", event, gid);
                }

                break;
            }

                /* Distribute soldier
                 * D */
            case 'D':
            {
                sscanf(buff, "%c", &event);
                DPRINT("%c\n", event);

                if ( distribute() ) {
                    DPRINT("%c succeeded\n", event);
                } else {
                    fprintf(stderr, "%c failed\n", event);
                }

                break;
            }

                /* General resigns from battle
                 * M <gid1> <gid2> */
            case 'M':
            {
                int gid1, gid2;

                sscanf(buff, "%c %d %d", &event, &gid1, &gid2);
                DPRINT("%c %d %d\n", event, gid1, gid2);

                if ( general_resign(gid1, gid2) ) {
                    DPRINT("%c %d %d succeeded\n", event, gid1, gid2);
                } else {
                    fprintf(stderr, "%c %d %d failed\n", event, gid1, gid2);
                }

                break;
            }

                /* Prepare for battle
                 * P <gid1> <gid2> */
            case 'P':
            {
                int gid1, gid2;
                sscanf(buff, "%c %d %d", &event, &gid1, &gid2);
                DPRINT("%c %d %d\n", event, gid1, gid2);

                if ( prepare_battle(gid1, gid2) ) {
                    DPRINT("%c %d %d succeeded\n", event, gid1, gid2);
                } else {
                    fprintf(stderr, "%c %d %d failed\n", event, gid1, gid2);
                }

                break;
            }

                /* Battle
                 * B <god_favor> <bit_stream> */
            case 'B':
            {
                int god_favor;
                char bit_stream[9];

                sscanf(buff, "%c %d %s\n", &event, &god_favor, bit_stream);
                DPRINT("%c %d %s\n", event, god_favor, bit_stream);

                if ( battle(god_favor, bit_stream, 8) ) {
                    DPRINT("%c %d %s succeeded\n", event, god_favor, bit_stream);
                } else {
                    fprintf(stderr, "%c %d %s failed\n", event, god_favor, bit_stream);
                }

                break;
            }

                /* Cease fire
                 * U */
            case 'U':
            {
                sscanf(buff, "%c", &event);
                DPRINT("%c\n", event);

                if ( cease_fire() ) {
                    DPRINT("%c succeeded\n", event);
                } else {
                    fprintf(stderr, "%c failed\n", event);
                }

                break;
            }

                /* Print general's soldiers
                 * W <gid> */
            case 'W':
            {
                int gid;

                sscanf(buff, "%c %d", &event, &gid);
                DPRINT("%c %d\n", event, gid);

                if ( print_generals_soldiers(gid) ) {
                    DPRINT("%c %d succeeded\n", event, gid);
                } else {
                    fprintf(stderr, "%c %d failed\n", event, gid);
                }

                break;
            }

                /* Print generals
                 * X */
            case 'X':
            {
                sscanf(buff, "%c", &event);
                DPRINT("%c\n", event);

                if ( print_generals() ) {
                    DPRINT("%c succeeded\n", event);
                } else {
                    fprintf(stderr, "%c failed\n", event);
                }

                break;
            }

                /* Print registration hashtable
                 * Y */
            case 'Y':
            {
                sscanf(buff, "%c", &event);
                DPRINT("%c\n", event);

                if ( print_registration_hashtable() ) {
                    DPRINT("%c succeeded\n", event);
                } else {
                    fprintf(stderr, "%c failed\n", event);
                }

                break;
            }

                /* Empty line */
            case '\n':
            break;

                /* Ignore everything else */
            default:
            DPRINT("Ignoring buff: %s \n", buff);

            break;
        }
    }

    free_world();
    return (EXIT_SUCCESS);
}
